### ISSUE TITLE
Computations in `begin..end` not evaluated for class/instance variables

### ISSUE THREAD

**Issue by devnote-dev** [crystal-lang/crystal#13766]
The following code fails to compile because the type of `@bar` cannot be inferred, despite the "return type" being an `Int32` (or `Int` at the very least) from `x += 1`:

```crystal
struct Foo
  @bar = begin
    x = 0
    x += 1
  end
  
  def bar
    @bar
  end
end
```

**Comment by straight-shoota** [comment]
I don't think this is a bug. It's a feature request for enhancing the compiler's type inference.

The `begin`/`end` block is not actually that relevant. A similar implementation without that cannot be inferred either:

```cr
struct Foo
  def initialize
    x = 0
    x += 1
    @bar = x # Error: can't infer the type of instance variable '@bar' of Foo
  end
end
 
Foo.new
```

The type inferrence could certainly be smarter about some things. It's often not trivial though. Maybe this case could be easy, I don't  know.

Applying the type restriction from a getter method seems obvious, but it's not that easy. In order to know that the return type of that method is associated to the type of an instance variable requires semantical analysis of the method. And that only happens if it's called. So it could be a bit flaky. Also it works only for trivial getters like this one where accesing the ivar is the only (or at least the last) expression.
In that case, it would be much easier to use the getter macro (`getter bar : Int32`) which actually applies the type restriction to the ivar.
Otherwise, it's an easy workaround to follow the compiler's suggestion and just add an explicit type restriction.
